activesupport/lib/active_support/deprecation/behaviors.rb

```diff
@@ -32,12 +32,16 @@ class Deprecation
         logger.debug callstack.join("\n  ") if debug
       },
 
-      notify: ->(message, callstack) {
-        ActiveSupport::Notifications.instrument("deprecation.rails",
-                                                message: message, callstack: callstack)
+      notify: ->(message, callstack, deprecation_horizon, gem_name) {
+        notification_name = "deprecation.#{gem_name.underscore.tr('/', '_')}"
```

Comment:
Sorry for digging this up after one year, adding the gem_name and the deprecation_horizon to the event is a great add however generating dynamically the notification name has one major disadvantage because that means we need to know in advance which gems have deprecations in order to keep track of them.

```ruby
class MySubscriber < ActiveSupport::Subscriber
  attach_to :rails
  # Prior to this PR, any gems using AS deprecations could be tracked thanks to this subscriber.
  # Now we can't because we'd need to attach the subscriber to other namespace which we don't know, thus the concept of "notifying" when a deprecation occur (no matter inside which gem) no longer exists

  def deprecation(event)
    ...
  end
end
```

My suggestion would be to use a static namespace like before (maybe change the `rails` namespace because it's not super accurate since any gems can trigger deprecation)

Reply:
I was not a big fan of this, but this was required for backwards compatibility.

I think you can use regular expressions for subscriptions, at least with the lower level API `ActiveSupport::Notifications::Fanout` API. That may not be propagated to `ActiveSupport::Subscriber` though.

Reply:
Ah ok I see thanks for the explanation, and TIL indeed the lower level API Fanout can be passed a regex. I'm going to open a PR to make it possible for `AS::Subscriber` as well

Reply:
> we need to know in advance which gems have deprecations in order to keep track of them.

That was exactly the reason why we did this. We should not thread deprecation in Rails in the same way we treat deprecation in other gems. At the time of a Rails upgrade, we don't care if the gem X also has deprecations, only if Rails has deprecation. Because of this we need to have different subscriber to each gem so we can chose the behavior we want based on the gem.